### PR TITLE
[IMP] runbot_merge: sign off commits by reviewer

### PR DESCRIPTION
**Pull request by KangOl**

### CODE REVIEW

File: runbot_merge/models/pull_requests.py [hunk 1]
```diff
@@ -1199,9 +1205,13 @@ def build_message(message, pr):
...
                 message)
-            if m:
-                return message
-            return message + '\n\ncloses {pr.repository.name}#{pr.number}'.format(pr=pr)
+            if not m:
+                message += '\n\ncloses {pr.repository.name}#{pr.number}'.format(pr=pr)
+            if pr.reviewed_by:
+                email = pr.reviewed_by.email or '%s@users.noreply.github.com' % pr.reviewed_by.github_login
+                reviewer = formataddr((pr.reviewed_by.name, email))
+                message += '\n\n\nSigned-off-by: {}'.format(reviewer)
```
Comment (xmo-odoo) [inline runbot_merge/models/pull_requests.py]: `commit -s` seems to only use two newlines e.g.
```
commit ccfb6bc700100d6bb108c2c0506ee2f448de2e7f (HEAD -> master-qweb-view-xmo)
Author: XXXX
Date:   Thu Nov 22 09:53:15 2018 +0100

    m
    
    Signed-off-by: XXXX
```

File: runbot_merge/tests/test_basic.py [hunk 2]
```diff
@@ -179,7 +183,31 @@ def test_commit_wrong_number(self, env, repo, users):
 
         master = repo.commit('heads/master')
         # closes on another repositoy, should modify the commit message
-        assert master.message == "simple commit message that closes #11\n\ncloses {repo.name}#1".format(repo=repo)
+        assert master.message == "simple commit message that closes #11\n\ncloses {repo.name}#1"\
+                                 "\n\n\nSigned-off-by: Reviewer <reviewer@example.com>".format(repo=repo)
+
+    def test_commit_delegate(self, env, repo, users):
+        """ verify 'closes ...' is correctly added in the commit message
+        """
```
Comment (xmo-odoo) [inline runbot_merge/tests/test_basic.py]: That doesn't seem like the correct docstring

File: runbot_merge/models/pull_requests.py (hunk 1 above)
Comment (KangOl) [reply runbot_merge/models/pull_requests.py]: Actually, github doesn't (currently) process it (see KangOl/testrepo@f9bb19 as example) and show commit message verbatim.
OTOH, `tig` will happily process this line regardless of surrounding  lines (empty or not)

But this "tag" is still misplaced as it interfere with [Co-authored-by:](https://help.github.com/articles/creating-a-commit-with-multiple-authors/) "tag" as it should be at the end of the commit message.

File: runbot_merge/tests/test_basic.py (hunk 2 above)
Comment (KangOl) [reply runbot_merge/tests/test_basic.py]: You're right. copy/paste error.

File: runbot_merge/models/pull_requests.py (hunk 1 above)
Comment (xmo-odoo) [reply runbot_merge/models/pull_requests.py]: > Actually, github doesn't (currently) process it (see KangOl/testrepo@f9bb19 as example) and show commit message verbatim.

I'm not saying anything about gh processing, just that `commit -s` uses two leading newlines not 3.

**Comment by KangOl** [comment]
recent example of "Co-authored-by" trailer that have been fucked-up: https://github.com/odoo/odoo/commit/02ce0022b2a5562a4b1c6f912a901d5b863df238

**Comment by KangOl** [comment]
Code rebased.

**Comment by xmo-odoo** [comment]
FWIW github actual still fails in test_pr_rebase_merge (and probably others but that's the first one which failed in my test run) because the PR assumes the email of reviewer is reviewer@example.com
```
E         Full diff:
E         - ('force rebuild\n\nuniquifier: EG8jrs0/MwBqL1d0',
E         + (~^force rebuild~,
E         frozenset({('title\n'
E         '\n'
E         'body\n'
E         '\n'
E         'closes odoo-mergebot-testing-org/repo#1\n'
E         '\n'
E         -              'Signed-off-by: Reviewer '
E         +              'Signed-off-by: Reviewer <reviewer@example.com>',
E         ?                                       ++++++++++++++++++++++ +
E         -              '<odoomergebotsacrifice1@users.noreply.github.com>',
E         frozenset({('B1',
E         frozenset({('B0',
E         frozenset({('M2',
E         frozenset({('M1',
E         frozenset({('M0',
E         frozenset())}))}))}))})),
E         ('M2',
E         frozenset({('M1',
E         frozenset({('M0', frozenset())}))}))}))}))
```

**Comment by KangOl** [comment]
I don't get how you end with the partner named "Reviewer". Where is it set?

**Comment by xmo-odoo** [comment]
> I don't get how you end with the partner named "Reviewer". Where is it set?

https://github.com/odoo/runbot/blob/master/runbot_merge/tests/local.py#L75

But that's when running the tests locally, when running them remotely you get what's defined for the role in your config file: https://github.com/odoo/runbot/blob/master/runbot_merge/tests/remote.py#L190

**Comment by KangOl** [comment]
Tests should now pass when run remotely

**Comment by xmo-odoo** [comment]
I'll try to run the testsuite this evening.

**Comment by xmo-odoo** [comment]
Still doesn't work:
```
E       AssertionError: assert {'A\n\ncloses...1', 'a2', ...} == {'A\n\ncloses ...1', 'a2', ...}
E         Extra items in the left set:
E         'B\n\ncloses odoo-mergebot-testing-org/repo#2\n\nSigned-off-by: Reviewer <odoomergebotsacrifice1@users.noreply.github.com>'
E         'A\n\ncloses odoo-mergebot-testing-org/repo#1\n\nSigned-off-by: Reviewer <odoomergebotsacrifice1@users.noreply.github.com>'
E         'C\n\ncloses odoo-mergebot-testing-org/repo#3\n\nSigned-off-by: Reviewer <odoomergebotsacrifice1@users.noreply.github.com>'
E         Extra items in the right set:
E         'A\n\ncloses odoo-mergebot-testing-org/repo#1\n\nSigned-off-by: False'
E         'C\n\ncloses odoo-mergebot-testing-org/repo#3\n\nSigned-off-by: False'
E         'B\n\ncloses odoo-mergebot-testing-org/repo#2\n\nSigned-off-by: False'
```

**Comment by KangOl** [comment]
This will hopefully works with last commit

**Comment by KangOl** [comment]
Still waiting a successful test run on actual github.com?

**Comment by xmo-odoo** [comment]
> Still waiting a successful test run on actual github.com?

Yep. It seems there was at least one actual issue (the `def users()` would call to the wrong endpoint) but it's a bit hard to know if they're all fixed when I get no test run passing entirely. I'm in contact with gh's support to try and improve the situation.

**Comment by KangOl** [comment]
You're right.
`users` endpoint is not call correctly. The indirection via `__call__` assume we want to call a route under the `/repos/` path, which is wrong for this endpoint

**Comment by xmo-odoo** [comment]
Yeah, I fixed it locally and will probably push it eventually but I'm still trying to get the actual suite to pass.

**Comment by KangOl** [comment]
Mhe!
With my last commit, mock of github api is now broken. Working on it.

**Comment by KangOl** [comment]
tests corrected.

**Pull request closed without merging**